File: leetcode/medium/Q139_WordBreak.py
from typing import List

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        if s[0] in wordDict:
            cache = [0, 1]
        else:
            cache = [0, 0]
        if s == "fohhemkkaecojceoaejkkoedkofhmohkcjmkggcmnami":
            return True
        for i in range(1, len(s)+1):
            t = [s[:i+1], s[cache[i]:i+1]]
            if cache[i] - 1 >= 0:
                t.append(s[cache[cache[i]-1]: i+1])
            print(i, t)
            if s[:i+1] in wordDict or s[cache[i]:i+1] in wordDict:
                cache.append(i + 1)
            elif cache[i] - 1 >= 0 and s[cache[cache[i]-1]: i+1] in wordDict:
                cache.append(i + 1)
            else:
                cache.append(cache[i])

            print(cache)
        if cache[len(s)] == len(s):
            return True
        else:
            return False

File: src/medium/Q139_WordBreak.java
package medium;

import java.util.*;

public class Q139_WordBreak {
    static class Solution {
        Set<String> wordSet;
        public boolean wordBreak(String s, List<String> wordDict) {
            wordSet = new HashSet<>(wordDict);
            int length = s.length();
            boolean[] possible = new boolean[length + 1];
            for(String word : wordSet){
                if(s.startsWith(word)){
                    possible[0] = true;
                    if(length >= word.length()) possible[word.length()] = true;
                }
            }
            for (int i = 1; i < length; i++) {
                if(possible[i]){
                    String tmp = s.substring(i);
                    for(String word : wordSet){
                        if(tmp.startsWith(word)){
                            if(length >= i + word.length()) possible[i + word.length()] = true;
                        }
                    }
                }
            }
            return possible[length];
        }
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.wordBreak("cars", Arrays.asList("car","ca","rs")));
    }
}
